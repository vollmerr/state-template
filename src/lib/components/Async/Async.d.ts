import * as React from 'react';

import * as Button from '../Button/Button';

export interface AsyncProps {
  /** Content to render */
  children: React.ReactNode;

  /** Error to render */
  error?: string;

  /** Action to remove the error */
  onDismiss?: () => void;

  /** Props to pass to `Button` in error message */
  btnProps?: Button.ButtonProps;

  /** Render loading indicator if true */
  isLoading: number | boolean;

  /** Delay until showing the loading indicator */
  delay?: number;
}

declare class Async extends React.Component<AsyncProps, {}> {}

export default Async;
