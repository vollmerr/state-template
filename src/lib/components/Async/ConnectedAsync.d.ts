import * as React from 'react';

import Async from './Async';

declare class ConnectedAsync extends Async {}

export default ConnectedAsync;
