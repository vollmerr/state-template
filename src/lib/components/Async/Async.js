import React from 'react';
import T from 'prop-types';

import ErrorMessage from './ErrorMessage';
import LoadingIndicator from './LoadingIndicator';

/**
 * Async sections have a loading indicator if they take too long
 * to load, and display an error message for errors
 */
const Async = (props) => {
  const {
    children,
    error,
    onDismiss,
    btnProps,
    isLoading,
    delay,
  } = props;

  return (
    <ErrorMessage error={error} onDismiss={onDismiss} btnProps={btnProps}>
      <LoadingIndicator isLoading={isLoading} delay={delay}>
        {children}
      </LoadingIndicator>
    </ErrorMessage>
  );
};

Async.propTypes = {
  /** Props to pass to `Button` in error message */
  btnProps: T.object,

  /** Content to render */
  children: T.node,

  /** Delay until showing the loading indicator */
  delay: T.number,

  /** Error to render */
  error: T.string,

  /** Render loading indicator if true */
  isLoading: T.oneOfType([
    T.number,
    T.bool,
  ]),

  /** Action to remove the error */
  onDismiss: T.func,
};

Async.defaultProps = {
  btnProps: undefined,
  children: null,
  delay: 200,
  error: null,
  isLoading: false,
  onDismiss: () => {},
};

export default Async;
