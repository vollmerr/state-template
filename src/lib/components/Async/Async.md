```jsx
import { ButtonRow, Button } from 'state-template';

class ExampleAsync extends React.Component {
  constructor(props) {
    super(props);
    this.state = { 
      error: null,
      hasErrorBtn: false,
      isLoading: false, 
      delay: 200,
    };
    // binding due to compilier not being able to use arrow functions...
    this.toggleError = this.toggleError.bind(this);
    this.toggleErrorBtn = this.toggleErrorBtn.bind(this);
    this.toggleLoading = this.toggleLoading.bind(this);
    this.increaseDelay = this.increaseDelay.bind(this);
    this.decreaseDelay = this.decreaseDelay.bind(this);
  }

  toggleError() {
    const { error } = this.state;
    this.setState({ error: error ? null : 'Example Errorz...' });
  }
      
  toggleErrorBtn() {
    const { hasErrorBtn } = this.state;
    this.setState({ hasErrorBtn: !hasErrorBtn });
  }

  toggleLoading() { 
    const { isLoading } = this.state;
    this.setState({ isLoading: !isLoading });
  }

  increaseDelay() { 
    const { delay } = this.state;
    this.setState({ delay: delay + 100 });
  }

  decreaseDelay() { 
    const { delay } = this.state;
    this.setState({ delay: delay - 100 });
  }

  render() {
    const { error, hasErrorBtn, isLoading, delay } = this.state;

    let btnProps;
    if (hasErrorBtn) {
      btnProps = {        
        text: 'Clear Error',
        variant: 'primary',
        onClick: () => alert('clicked me!'),
      };
    }

    return (
      <>
        <ButtonRow>
          <Button text={'Toggle loading'} onClick={this.toggleLoading} />
          <Button text={'Increase delay'} onClick={this.increaseDelay} />
          <Button text={'Decrease delay'} onClick={this.decreaseDelay} />
          <Button text={'Toggle Error'} onClick={this.toggleError} />
          <Button text={'Toggle Error Button'} onClick={this.toggleErrorBtn} active={hasErrorBtn} />
        </ButtonRow>
        <p>Delay: {delay}ms</p>
        
        <Async 
          error={error} 
          onDismiss={this.toggleError} 
          btnProps={btnProps} 
          isLoading={isLoading} 
          delay={delay}
        >
          <p>Content goes here...</p>
        </Async>
      </>
    );
  }
}

<ExampleAsync />
```
