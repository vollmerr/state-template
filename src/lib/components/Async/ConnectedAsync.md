```jsx
  const { connect } = require('react-redux');
  const { asyncActions } = require('state-template');
  
  // example redux initial store
  const store = { 
    async: {
      isLoading: 0,
      error: null,
    },
  };

  class ExampleAsync extends React.Component {
    constructor(props) {
      super(props);
      this.state = { 
        hasErrorBtn: false,
        delay: 200,
      };
      // binding due to compilier not being able to use arrow functions...
      this.toggleError = this.toggleError.bind(this);
      this.toggleErrorBtn = this.toggleErrorBtn.bind(this);
      this.toggleLoading = this.toggleLoading.bind(this);
      this.increaseDelay = this.increaseDelay.bind(this);
      this.decreaseDelay = this.decreaseDelay.bind(this);
    }

    toggleError() {
      const { error, setError } = this.props;
      setError(error ? null : 'Example Errorz...');
    }
        
    toggleErrorBtn() {
      const { hasErrorBtn } = this.state;
      this.setState({ hasErrorBtn: !hasErrorBtn });
    }

    toggleLoading() { 
      const { isLoading, increaseLoading, decreaseLoading } = this.props;

      if (isLoading) {
        decreaseLoading();
      } else {
        increaseLoading();
      }
    }

    increaseDelay() { 
      const { delay } = this.state;
      this.setState({ delay: delay + 100 });
    }

    decreaseDelay() { 
      const { delay } = this.state;
      this.setState({ delay: delay - 100 });
    }

    render() {
      const { hasErrorBtn, delay } = this.state;

      let btnProps;
      if (hasErrorBtn) {
        btnProps = {        
          text: 'Clear Error',
          variant: 'primary',
          onClick: () => alert('clicked me!'),
        };
      }

      return (
        <>
          <ButtonRow>
            <Button text={'Toggle loading'} onClick={this.toggleLoading} />
            <Button text={'Increase delay'} onClick={this.increaseDelay} />
            <Button text={'Decrease delay'} onClick={this.decreaseDelay} />
            <Button text={'Toggle Error'} onClick={this.toggleError} />
            <Button text={'Toggle Error Button'} onClick={this.toggleErrorBtn} active={hasErrorBtn} />
          </ButtonRow>
          <p>Delay: {delay}ms</p>
          
          <ConnectedAsync delay={delay} btnProps={btnProps}>
            <p>Content goes here...</p>
          </ConnectedAsync>
        </>
      );
    }
  }

  const mapStateToProps = state => ({
    isLoading: state.async.isLoading,
    error: state.async.error,
  })

  const mapDispatchToProps = dispatch => ({
    setError: props => dispatch(asyncActions.setError(props)),
    increaseLoading: () => dispatch(asyncActions.increaseLoading()),
    decreaseLoading: () => dispatch(asyncActions.decreaseLoading()),
  });

  const WithRedux = connect(mapStateToProps, mapDispatchToProps)(ExampleAsync);

  <ExampleRedux store={store}>
    <WithRedux />
  </ExampleRedux>
```
