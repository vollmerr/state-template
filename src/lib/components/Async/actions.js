import { createAction } from 'redux-actions';

import * as C from './constants';

export const setError = createAction(C.SET_ERROR);
export const increaseLoading = createAction(C.INCREASE_LOADING);
export const decreaseLoading = createAction(C.DECREASE_LOADING);
