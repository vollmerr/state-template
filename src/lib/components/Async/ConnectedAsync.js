import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import Async from './Async';
import * as actions from './actions';
import * as selectors from './selectors';

export const mapStateToProps = createStructuredSelector({
  isLoading: selectors.getIsLoading(),
  error: selectors.getError(),
});

export const mapDispatchToProps = dispatch => ({
  onDismiss: () => dispatch(actions.setError(null)),
});

const ConnectedAsync = connect(mapStateToProps, mapDispatchToProps)(Async);

export default ConnectedAsync;
