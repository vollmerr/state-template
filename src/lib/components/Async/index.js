import * as asyncActions from './actions';
import * as asyncConstants from './constants';
import asyncReducer from './reducer';
import * as asyncSelectors from './selectors';

export {
  asyncActions,
  asyncConstants,
  asyncReducer,
  asyncSelectors,
};

export { default } from './Async';
export { default as ConnectedAsync } from './ConnectedAsync';
export { default as ErrorMessage } from './ErrorMessage';
export { default as LoadingIndicator } from './LoadingIndicator';
