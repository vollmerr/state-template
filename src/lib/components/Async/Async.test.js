import React from 'react';
import { shallow } from 'enzyme';

import { Async as fromLib } from '../..';
import Async from '.';

const props = {
  error: 'test-error',
  onDismiss: jest.fn(),
  btnProps: { onClick: jest.fn(), text: 'test-button' },
  isLoading: false,
  delay: 1000,
};

let wrapper;
describe('Async', () => {
  beforeEach(() => {
    wrapper = shallow(<Async {...props}>children...</Async>);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(Async);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
