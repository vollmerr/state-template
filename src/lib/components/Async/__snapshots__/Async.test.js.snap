// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async should render correctly 1`] = `
<ErrorMessage
  btnProps={
    Object {
      "onClick": [MockFunction],
      "text": "test-button",
    }
  }
  error="test-error"
  onDismiss={[MockFunction]}
>
  <LoadingIndicator
    delay={1000}
    isLoading={false}
  >
    children...
  </LoadingIndicator>
</ErrorMessage>
`;
